server/src/main/java/io/kroki/server/service/Plantuml.java

```diff
@@ -153,30 +229,73 @@ static String withDelimiter(String source) {
+        // no path walking
+        && !include.startsWith("../") && !include.contains("/../") && !include.endsWith("/..")
+        && !include.startsWith("..\\") && !include.contains("\\..\\") && !include.endsWith("\\..")
+        && !include.contains("/..\\") && !include.contains("\\../")) {
+        // PlantUML's safety checking will suffice as it will only allow files directly in the include path
+
+        // Note: we are relying on the PlantUML include checking algorithm:
+        // * See ImportedFiles#executeInclude and ImportedFiles#executeIncludesub
+        // * !importsub does not work with URLs or the standard library
+        // * Imports that start with `http://` or `https://` are resolved from the URL, that is generally
+        //   unsafe unless you can trust the source of the URL, thus we do not permit here.
+        // * Imports wrapped in `<` and `>` are resolved only from the standard library and thus should be
+        //   a fixed set guarded by `/stdlib/${name}-abx.repx` resources on the classpath, in any case we
+        //   ignore stdlib import validation here
+        // * Absolute imports could refer to any file, so we exclude those, though FileWithSuffix#fileOk()
+        //   should restrict accessible files to only those "approved"
+        // * We only auto-permit "search path" includes that do not attempt search path escaping with `..`
+        // * The import or importsub that we auto-permit will result in a call to ImportedFiles.getFile(...)
+        // * That will iterate the "plantuml.include.path" (note PlantUML is not well designed for embedding
+        //   on this property as the ImportedFiles#INCLUDE_PATH is initialized once on classloading)
+        // * At this point, if the INCLUDE_PATH included say `/example` and the include was `foo/bar.puml`
+        //   and there is a file `/example/foo/bar.puml` then that would stop the search with an `AFileRegular`
+        //   instance. HOWEVER, the return value of ImportedFiles#getFile is guarded by ImportedFiles#isAllowed
+        // * ImportedFiles#isAllowed will only permit AFile instances with the AFile#getSystemFolder() being
+        //   contained in ImportedFiles#INCLUDE_PATH so as AFileRegular#getSystemFolder() always returned the
+        //   parent folder of the file and include of `foo/bar.puml` found on an include path of `/example`
+        //   resolving to `/example/foo/bar.puml` will have a system folder of `/example/foo` which is not
+        //   one of the folders whitelisted in the include path and thus the include will be resolved as not-found
+        //
+        // Effectively only imports that are immediate children of the folders listed in "plantuml.include.path"
+        // are eligible for inlcude (unless OptionFlags.ALLOW_INCLUDE has been set to true)
+        sb.append(line).append("\n");
+      } else if (includeWhitelist.stream().anyMatch(p -> p.matcher(include).matches())) {
```

Comment:
Not sure which parameter should have the priority between the safe mode and the include whitelist options. I am somewhat leaning toward safe mode. In other words, I think that if safe mode SECURE is defined we should ignore whitelist. Arguably it's a bit restrictive but if it's clearly stated in the documentation that's OK.
Plus, if we do the opposite you cannot have this behavior.

Reply:
I went ahead and now safe mode SECURE has the priority over the whitelist.